File: algorithms/src/labs/lab_3/G.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.PrintWriter;
import java.util.Random;
import java.util.StringTokenizer;

/**
 * Created by woofi on 20.12.2015.
 */
public class G {
    private static final Random rand = new Random();

    static class ImplicitTreap {
        static class PairNode {
            Node left;
            Node right;

            PairNode(Node l, Node r) {
                left = l;
                right = r;
            }
        }

        static class Node {
            long len; // river
            long square; // river

            int y;
            int size;
            Node left;
            Node right;

            Node(long len, int y, Node left, Node right) {
                this.len = len; // river
                this.y = y;
                this.left = left;
                this.right = right;
                fix();
            }

            private int size(Node node) {
                return (node != null) ? node.size : 0;
            }

            // river
            private long square(Node node) {
                return (node != null) ? node.square : 0;
            }

            private void fix() {
                square = square(left) + square(right) + len * len; // river
                size = size(left) + size(right) + 1;
            }
        }

        Node root;
        int size;

        ImplicitTreap() {
            root = null;
            size = 0;
        }

        private int size(Node node) {
            return (node != null) ? node.size : 0;
        }

        private Node merge(Node l, Node r) {
            if (l == null) {
                return r;
            }
            if (r == null) {
                return l;
            }

            if (l.y > r.y) {
                return new Node(l.len, l.y, l.left, merge(l.right, r));
            } else {
                return new Node(r.len, r.y, merge(l, r.left), r.right);
            }
        }

        private PairNode split(Node pivot, int x) {
            PairNode pn = new PairNode(null, null);
            PairNode t = new PairNode(null, null);
            int pivotIndex = size(pivot.left) + 1;

            if (pivotIndex <= x) {
                if (pivot.right == null) {
                    pn.right = null;
                } else {
                    t = split(pivot.right, x - pivotIndex);
                    pn.right = t.right;
                }
                pn.left = new Node(pivot.len, pivot.y, pivot.left, t.left);
            } else {
                if (pivot.left == null) {
                    pn.left = null;
                } else {
                    t = split(pivot.left, x);
                    pn.left = t.left;
                }
                pn.right = new Node(pivot.len, pivot.y, t.right, pivot.right);
            }

            return pn;
        }

        void add(int x, long len) {
            if (root == null) {
                root = new Node(len, rand.nextInt(), null, null);
            } else {
                PairNode pn = split(root, x);
                root = merge(merge(pn.left, new Node(len, rand.nextInt(), null, null)), pn.right);
            }

            //System.out.println(x + " " + len + " root: " + root.len + " " + root.y + " " + root.size);
            size++;
        }

        long remove(int x) {
            long len = 0;

            if (root != null) {
                Node t;
                PairNode pn = split(root, x);
                t = pn.left;
                pn = split(pn.right, 1);
                root = merge(t, pn.right);
                len = pn.left.len;
            }

            size--;
            return len; // self
        }

        /* ------ RIVER ------ */

        void separate(int x) {
            long len = remove(x);
            add(x, len >> 1);
            add(x + 1, (len >> 1) + len % 2);
        }

        void bankrupt(int x) {
            if (x == 0) {
                long lenFirst = remove(0);
                long lenSecond = remove(0);
                add(0, lenFirst + lenSecond);
            } else if (x == size - 1) {
                long lenLast = remove(size - 1);
                long lenPenult = remove(size - 1);
                add(size, lenPenult + lenLast);
            } else {
                long lenBefore = remove(x - 1);
                long len = remove(x - 1);
                long lenAfter = remove(x - 1);
                add(x - 1, lenBefore + (len >> 1));
                add(x, (len >> 1) + len % 2 + lenAfter);
            }
        }

        long getAnswer() {
            return root.square;
        }
    }

    public static void main(String args[]) throws Exception {
        Scanner sc = new Scanner(new File("river.in"));
        PrintWriter pw = new PrintWriter(new File("river.out"));
        ImplicitTreap it = new ImplicitTreap();

        int n = sc.nextInt();
        sc.nextInt();
        for (int i = 0; i < n; i++) {
            it.add(i, sc.nextLong());
        }
        pw.println(it.getAnswer());

        int m = sc.nextInt();
        while (m --> 0) {
            switch (sc.nextInt()) {
                case 1:
                    it.bankrupt(sc.nextInt() - 1);
                    pw.println(it.getAnswer());
                    break;
                case 2:
                    it.separate(sc.nextInt() - 1);
                    pw.println(it.getAnswer());
                    break;
            }
        }

        sc.close();
        pw.close();
    }

    static class Scanner {
        BufferedReader br;
        StringTokenizer t;

        Scanner(File file) throws Exception {
            br = new BufferedReader(new FileReader(file));
            t = new StringTokenizer("");
        }

        boolean hasNext() throws Exception {
            while (!t.hasMoreTokens()) {
                String line = br.readLine();
                if (line == null)
                    return false;
                t = new StringTokenizer(line);
            }
            return true;
        }

        String next() throws Exception {
            if (hasNext()) {
                return t.nextToken();
            }
            return null;
        }

        int nextInt() throws Exception {
            return Integer.parseInt(next());
        }

        long nextLong() throws Exception {
            return Long.parseLong(next());
        }

        double nextDouble() throws Exception {
            return Double.parseDouble(next());
        }

        void close() throws Exception {
            br.close();
        }
    }
}

File: algorithms/labs/lab-8/G.kt
/**
 * Nariman Safiulin (woofilee)
 * File: G.kt
 * Created on: Oct 23, 2016
 */

import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.PrintWriter
import java.util.*

private val PROBLEM_NAME = "carno"

private class Scanner(file: File) {
    val br = BufferedReader(FileReader(file))
    var st = StringTokenizer("")

    fun hasNext(): Boolean {
        while (!st.hasMoreTokens()) {
            val line = br.readLine() ?: return false
            st = StringTokenizer(line)
        }
        return true
    }

    fun next(): String = if (hasNext()) st.nextToken() else ""
    fun nextInt(): Int = Integer.parseInt(next())
    fun nextLong(): Long = java.lang.Long.parseLong(next())
    fun nextDouble(): Double = java.lang.Double.parseDouble(next())
    fun close() = br.close()
}

private fun solve(`in`: Scanner, out: PrintWriter) {
    fun power(e: Int) = 1 shl e
    fun log(v: Int): Int = (Math.log(v.toDouble()) / Math.log(2.0)).toInt() + 1

    val max = 2 * 100000
    var n = 0
    val p = Array(max) { 0 }
    val d = Array(max) { 0 }
    val dp = Array(max) { Array(log(max) + 1) { 0 } }
    val relation = Array(max) { it }

    fun lca(v: Int, u: Int): Int {
        if (d[v] > d[u]) return lca(u, v)

        var v = v
        var u = u

        (log(n) downTo 0).forEach {
            if (d[u] - d[v] >= power(it))
                u = dp[u][it]
        }

        if (v == u) return v

        (log(n) downTo 0).forEach {
            if (dp[v][it] != dp[u][it]) {
                v = dp[v][it]
                u = dp[u][it]
            }
        }

        return p[v]
    }

    fun find(dino: Int): Int = if (dino == relation[dino]) dino else { // I want a tailrec! :(
        relation[dino] = find(relation[dino])
        relation[dino]
    }

    (1..`in`.nextInt()).forEach {
        when (`in`.next()) {
            "+" -> {
                n++

                val dino = `in`.nextInt() - 1

                p[n] = dino
                d[n] = d[dino] + 1
                dp[n][0] = dino

                (1..log(n)).forEach { i ->
                    dp[n][i] = dp[dp[n][i - 1]][i - 1]
                }
            }
            "-" -> {
                val dino = `in`.nextInt() - 1
                relation[dino] = dp[dino][0]
            }
            "?" -> out.println(find(lca(`in`.nextInt() - 1, `in`.nextInt() - 1)) + 1)
        }
    }
}

fun main(args: Array<String>) {
    val `in` = Scanner(File(PROBLEM_NAME + ".in"))
    val out = PrintWriter(File(PROBLEM_NAME + ".out"))

    solve(`in`, out)

    `in`.close()
    out.close()
}
